// parms/Gruntfile.js
module.exports = function(grunt)
{
	grunt.initConfig({
		pkg: grunt.file.readJSON('package.json'),
		jshint: {
			all: {
				options: {
					jshintrc: true
				},
				src: ['Gruntfile.js', 'src/**/*.js', 'test/**/*.js']
			}
		},
		mochaTest: {
			all: {
				options: {
					reporter: 'spec'
				},
				src: ['test/**/*.js']
			}
		},
		yuidoc: {
			all: {
				name: '<%= pkg.name %>',
				description: '<%= pkg.description %>',
				version: '<%= pkg.version %>',
				url: '<%= pkg.homepage %>',
				options: {
					paths: 'src/.',
					outdir: 'docs'
				}
			}
		},
		browserify: {
			all: {
				options: {
					bundleOptions: {
						standalone: 'parms'
					}
				},
				files: {
					"dist/parms.js": ["src/parms.js"]
				}
			}
		},
		uglify: {
			all: {
				options: {
					mangle: true,
					compress: true
				},
				files: {
					"dist/parms.min.js": "dist/parms.js"
				}
			}
		}
	});

	grunt.loadNpmTasks('grunt-contrib-jshint');
	grunt.loadNpmTasks('grunt-mocha-test');
	grunt.loadNpmTasks('grunt-contrib-yuidoc');
	grunt.loadNpmTasks('grunt-browserify');
	grunt.loadNpmTasks('grunt-contrib-uglify');

	grunt.registerTask('default', [
		'jshint',
		'mochaTest',
		'yuidoc',
		'browserify',
		'uglify'
	]);
};
